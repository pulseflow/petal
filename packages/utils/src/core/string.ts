import { isObject } from './assert';

/**
 * Replace backslash to slash
 *
 * @category String
 */
export function slash(str: string) {
	return str.replace(/\\/g, '/');
}

/**
 * Ensure prefix of a string
 *
 * @category String
 */
export function ensurePrefix(prefix: string, str: string) {
	if (!str.startsWith(prefix))
		return prefix + str;
	return str;
}

/**
 * Ensure suffix of a string
 *
 * @category String
 */
export function ensureSuffix(suffix: string, str: string) {
	if (!str.endsWith(suffix))
		return str + suffix;
	return str;
}

/**
 * Dead simple template engine, just like Python's `.format()`
 * Support passing variables as either in index based or object/name based approach
 * While using object/name based approach, you can pass a fallback value as the third argument
 *
 * @category String
 * @example
 */
export function template(str: string, object: Record<string | number, any>, fallback?: string | ((key: string) => string)): string;
export function template(str: string, ...args: (string | number | bigint | undefined | null)[]): string;
export function template(str: string, ...args: any[]): string {
	const [firstArg, fallback] = args;

	if (isObject(firstArg)) {
		const vars = firstArg as Record<string, any>;
		return str.replace(/{([\w\d]+)}/g, (_, key) => vars[key] || ((typeof fallback === 'function' ? fallback(key) : fallback) ?? key));
	}
	else {
		return str.replace(/{(\d+)}/g, (_, key) => {
			const index = Number(key);
			if (Number.isNaN(index))
				return key;
			return args[index];
		});
	}
}

// port from nanoid
// https://github.com/ai/nanoid
const urlAlphabet = 'useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict';
/**
 * Generate a random string
 * @category String
 */
export function randomStr(size = 16, dict = urlAlphabet) {
	let id = '';
	let i = size;
	const len = dict.length;
	while (i--)
		id += dict[(Math.random() * len) | 0];
	return id;
}

/**
 * First letter uppercase, other lowercase (I HATE CAPITALISM)
 * @category string
 * @example
 * ```
 * capitalize('hello') => 'Hello'
 * ```
 */
export function capitalize(str: string): string {
	return str[0].toUpperCase() + str.slice(1).toLowerCase();
}

const _reFullWs = /^\s*$/;

/**
 * Remove common leading whitespace from a template string.
 * Will also remove empty lines at the beginning and end.
 * @category string
 * @example
 * ```ts
 * const str = unindent`
 *   if (a) {
 *     b()
 *   }
 * `
 */
export function unindent(str: TemplateStringsArray | string) {
	const lines = (typeof str === 'string' ? str : str[0]).split('\n');
	const whitespaceLines = lines.map(line => _reFullWs.test(line));

	const commonIndent = lines
		.reduce((min, line, idx) => {
			if (whitespaceLines[idx])
				return min;
			const indent = line.match(/^\s*/)?.[0].length;
			return indent === undefined ? min : Math.min(min, indent);
		}, Number.POSITIVE_INFINITY);

	let emptyLinesHead = 0;
	while (emptyLinesHead < lines.length && whitespaceLines[emptyLinesHead])
		emptyLinesHead++;
	let emptyLinesTail = 0;
	while (emptyLinesTail < lines.length && whitespaceLines[lines.length - emptyLinesTail - 1])
		emptyLinesTail++;

	return lines
		.slice(emptyLinesHead, lines.length - emptyLinesTail)
		.map(line => line.slice(commonIndent))
		.join('\n');
}
