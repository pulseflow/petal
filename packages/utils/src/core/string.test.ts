import { capitalize, ensurePrefix, ensureSuffix, slash, template, unindent } from './string';

it('template', () => {
	expect(
		template(
			'Hello {0}! My name is {1}.',
			'Paulinè',
			'Pauline',
		),
	).toEqual('Hello Paulinè! My name is Pauline.');

	expect(
		template(
			'{0} + {1} = {2}{3}',
			1,
			'1',
			// @ts-expect-error disallow non-literal on type
			{ v: 2 },
			[2, 3],
		),
	).toEqual('1 + 1 = [object Object]2,3');

	expect(
		template(
			'{10}',
		),
	).toEqual('undefined');

	expect(
		template(
			'Hi',
			'',
		),
	).toEqual('Hi');
});

it('namedTemplate', () => {
	expect(
		template(
			'{greet}! My name is {name}.',
			{ greet: 'Hello', name: 'Pauline' },
		),
	).toEqual('Hello! My name is Pauline.');

	expect(
		template(
			'{a} + {b} = {result}',
			{ a: 1, b: 2, result: 3 },
		),
	).toEqual('1 + 2 = 3');

	expect(
		template(
			'{1} + {b} = 3',
			{ 1: 'a', b: 2 },
		),
	).toEqual('a + 2 = 3');

	expect(
		template(
			'{10}',
			{},
		),
	).toEqual('10');

	expect(
		template(
			'{11}',
			null,
		),
	).toEqual('undefined');

	expect(
		template(
			'{11}',
			undefined,
		),
	).toEqual('undefined');

	expect(
		template(
			'{10}',
			{},
			'unknown',
		),
	).toEqual('unknown');

	expect(
		template(
			'{1} {2} {3} {4}',
			{ 4: 'known key' },
			k => String(+k * 2),
		),
	).toEqual('2 4 6 known key');
});

it('slash', () => {
	expect(slash('\\123')).toEqual('/123');
	expect(slash('\\\\')).toEqual('//');
	expect(slash('\\\h\\\i')).toEqual('/h/i');
});

it('ensurePrefix', () => {
	expect(ensurePrefix('abc', 'abcdef')).toEqual('abcdef');
	expect(ensurePrefix('hi ', 'jack')).toEqual('hi jack');
});

it('ensureSuffix', () => {
	expect(ensureSuffix('world', 'hello ')).toEqual('hello world');
	expect(ensureSuffix('123', 'abc123')).toEqual('abc123');
});

it('capitalize', () => {
	expect(capitalize('hello World')).toEqual('Hello world');
	expect(capitalize('123')).toEqual('123');
	expect(capitalize('中国')).toEqual('中国');
	expect(capitalize('āÁĂÀ')).toEqual('Āáăà');
	expect(capitalize('\a')).toEqual('A');
});

it('unindent', () => {
	expect(
		unindent`
      if (a) {
        b()
      }
    `,
	).toMatchSnapshot('base');

	expect(
		unindent`
          if (a) {
        b()
      }
    `,
	).toMatchSnapshot('with leading indent');

	expect(
		unindent`

          if (a) {
        b()
      }

    `,
	).toMatchSnapshot('multi-start and end');

	expect(
		unindent`if (a) {
  b()
}`,
	).toMatchSnapshot('no start or end');

	expect(
		unindent`
                  if (a) {
                  b()
              }
    `,
	).toMatchSnapshot('indent deep');
});
