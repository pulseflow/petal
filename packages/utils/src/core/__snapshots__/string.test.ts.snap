// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unindent > base 1`] = `
"if (a) {
  b()
}"
`;

exports[`unindent > indent deep 1`] = `
"    if (a) {
    b()
}"
`;

exports[`unindent > multi-start and end 1`] = `
"    if (a) {
  b()
}"
`;

exports[`unindent > no start or end 1`] = `
"if (a) {
  b()
}"
`;

exports[`unindent > with leading indent 1`] = `
"    if (a) {
  b()
}"
`;
